### ISSUE TITLE
"User X does not have appropriate auth credentials in kubeconfig" when listing namespaces in Visual Studio

### ISSUE THREAD

**Issue by daniv-msft** [microsoft/mindaro#26]
From what we've seen, this is related to enabling on an AKS cluster AAD managed authentication where multi-factor authentication is required. Potentially, this might also be caused by other AAD managed authentication settings.

Symptoms of the issue:
![image](https://user-images.githubusercontent.com/48293037/92552294-2c519300-f215-11ea-9e46-d76b315bb498.png)

As of today, Visual Studio pulls a temporary kubeconfig when configuring Local Process with Kubernetes. Depending on the AKS' cluster settings, it might be needed at this point to approve the device for login, but this scenario is not handled properly by Local Process with Kubernetes.
This explains why if you run a similar command in kubectl, it works: the device login was most likely already configured there.

* Short term workaround: please consider using the [VS Code extension](https://aka.ms/mindaro-marketplace-vscode). Because it doesn't pull any temporary kubeconfig (it only relies on the default one in `%UserProfile%/.kube/config`), if you're able to run `kubectl get namespaces` on your default kubeconfig, then you should be setup properly.
[How to use Local Process with Kubernetes in VS Code](https://aka.ms/localprocesswithk8s-vsc)

* Middle term workaround (not available yet): we're working to persist the kubeconfigs pulled in Visual Studio. That way, it would be possible to run the device login flow on it once and get unblocked.

* Long term solution: we're evaluating how we can handle this kind of auth issues without reimplementing the same authentication logic present in kubectl.

**Comment by danegsta** [comment]
With the 1.0 release of the VS extension, we've enabled a workaround for this particular error. If you get this error message while refreshing resources for your cluster with AAD managed authentication, you can follow these steps to get unblocked:

1. Ensure you have kubectl installed (either run `az aks install-cli` if you have the azure cli installed or follow the directions at https://kubernetes.io/docs/tasks/tools/install-kubectl/)
2. Find the cached kubeconfig the Bridge to Kubernetes VS extension is using to connect to your cluster: %LOCALAPPDATA%\BridgeToKubernetes\cache\<subscription_id>\<resourcegroup_name>\<cluster_name>.kubeconfig
3. Open a new console or powershell window
4. Set a temporary KUBECONFIG environment variable in your terminal pointing to the the kubeconfig file from step 2 (`set KUBECONFIG="<kubeconfig_path>"` for cmd or `$env:KUBECONFIG="<kubeconfig_path>"` for powershell)
5. Run a kubectl command from the terminal session such as `kubectl get namespaces`
6. Follow the instructions from kubectl for authenticating the cluster connection
7. Once the authenticaiton process is complete, return to VS and refresh the cluster namespaces (or relaunch the configuration dialog if you closed it) and you should be able to access the cluster resources

**Comment by irperez** [comment]
We're running into this now as we just integrated AAD authentication.

This is now a chore for our devs to work around, especially if they don't know K8s well. Just a nudge to get a better flow here 👍 

Loving LPK!

**Comment by daniv-msft** [comment]
Thank you @irperez! Yes, this issue is tricky on our side, as the final solution we consider has a lot of implications for us.
However it might be possible to help with the experience in product without going down the path of reimplementing kubectl.

**Comment by irperez** [comment]
It would be ideal to just use the same kubectl config that I use from the kubectl command.  Although I could see how that could cause conflicts in some situations.  Making this an option that I can set in VStudio might be easier for our devs to navigate.  Having them work with two .kubeconfig was not obvious at all.  Just throwing some ideas out there.

**Comment by danegsta** [comment]
@daniv-msft that's pretty much our current estimated timeline. I'll provide an update here once support for using clusters in the local kube config goes live in the Visual Studio extension.

**Comment by irperez** [comment]
We can no longer wait for the extension to get updated.  This bug hit us so hard that we decided to pull AAD Auth from the dev cluster and do a full rebuild.  The work around works temporarily.  After a while it times out and the developers get into a bad state with kubectl where they can no longer access the cluster.

**Comment by daniv-msft** [comment]
@irperez Sorry this bug is making your life harder. :( We're making good progress on reconciling the kubeconfig context into the main kubeconfig, but we're not ready to ship it yet.

However what you say about the work around working only temporarily concerns me. Could you please provide more information about the bad state with kubectl you're mentioning?
What is the status at this point: does kubectl offer a new flow to authenticate again or are you stuck with a non-actionable error?

@danegsta My understanding of this error was that the fix would be a one time action to authenticate the device. Would you have any other information about users having to do this regularly?

**Comment by irperez** [comment]
@daniv-msft I'm not totally sure what happened. But we did follow the work around by creating the environment variable.  That worked for several weeks.  It seems though that the token timed out.  Not sure.  Within the same week, 3 developers ran into a wall where kubectl was no longer able to authenticate.  We had to wipe out the KUBECONFIG environment variable.  And that restored kubectl, but broke visual studio's work around.

I will talk to our cluster admin to see how the AAD is configured against the cluster.  But the AAD itself is not administered by our cluster admin.  Its administered by other individuals that are not part of this project.

**Comment by daniv-msft** [comment]
Thanks @irperez for providing more information. I agree that if 3 developers ran into the same issue the same week, this sounds like either expiring tokens, or a change in AAD configuration that would have invalidated the existing credentials.

By any chance, would the developer in your team still have some data about this issue (error returned by kubectl, for example)? If kubectl was no longer able to authenticate, then getting the kubectl dev team's opinion about this would be helpful.
Even once we'll reconcile the kubeconfig context with the main kubeconfig, we'll still rely on kubectl for authentication in the foreseeable future to workaround this issue.

**Comment by davejhahn** [comment]
I'm not able to get this to work, not sure what the problem is... We are using integrated AAD, so went through the approach described by @danegsta and was able to get past the problem with namespaces not loading.  

Running in isolation mode , VS run as admin, confirmed it says "Admin" in upper right of Visual Studio, and when running "Bridge to Kubernetes" it prompts me with "Elevation Required", I select OK, and it hangs on "Connecting to cluster" and eventually fails with "Failed to get routing manager deployment status"

Tried the same not in isolation mode, and it was able to get further, saw connections hitting my local VS instance when expected, however started getting errors that it couldn't connect to redis.  I am guessing that is related to environment settings, will look at that further with the options available.  

But I definitely want to use isolated, it would be disruptive across the team if everything starts hitting an individual developer's VS instance while he/she happens to be working on it.

One caveat to our situation, we are using AGIC--does this cause any problems?

**Comment by davejhahn** [comment]
On a side note, does anyone know the actual retirement date of Azure Dev Spaces?  It says October 2023 https://docs.microsoft.com/en-us/azure/dev-spaces/migrate-to-bridge-to-kubernetes but we started getting notifications in Visual Studio that it will no longer be supported November 15th, 2021.

**Comment by daniv-msft** [comment]
@davejhahn Indeed, as of today using Bridge with AGIC requires more steps (sending the HTTP header manually). You can read more here: https://github.com/microsoft/mindaro/issues/50

However, I suspect that you're encountering a different issue before that, because the AGIC issue doesn't correspond with the error message you mention (`Failed to get routing manager deployment status`).
Would it be possible for you to send us (`BridgeToKubernetes@microsoft.com`) the logs from the routing manager pod running in your cluster, as well as the corresponding logs generated in:
For Windows: %TEMP%/Bridge to Kubernetes
For OSX/Linux: $TMPDIR/Bridge to Kubernetes

Regarding the retirement date of Azure Dev Spaces, I confirm that it is in October 2023. However, creating Azure Dev Spaces controllers from Visual Studio won't be supported anymore after 11/15/**2020** (and not 2021). If you need to create a controller after this date, you can still use the [Azure Dev Spaces CLI](https://docs.microsoft.com/azure/dev-spaces/how-to/install-dev-spaces).
However, we would be happy to help you transitioning to Bridge as we believe this is a better solution and user experience.
Could you please provide us with the logs mentioned above, so that we can help unblock you?

**Comment by daniv-msft** [comment]
If it helps, here is how to retrieve the routing manager logs:
1. Retrieve the routing manager pod name
2. kubectl -n {namespace} get pods
3. Retrieve logs for the routing manager pod
4. kubectl -n {namespace} logs {routing manager pod name from step 1} > routing-manager-logs.txt

**Comment by TomGeske** [comment]
Could Visual Studio Extension for Bridge to Kubernetes use the same .kube/config like other tools do? 
For auth workflow would integrating [kubelogin ](https://github.com/Azure/kubelogin) be an option?

@weinong

**Comment by amsoedal** [comment]
@TomGeske Hi Tom, I think as of [our latest release](https://github.com/microsoft/mindaro/issues/97#issue-774062290) the VS extension can connect to any cluster that's present in the local kubeconfig.  Are you seeing issues with that? 
Also I'll log a work item for the team to investigate kubelogin!
